server/src/test/java/org/elasticsearch/action/ActionListenerTests.java

```diff
@@ -370,6 +374,75 @@ public void onFailure(Exception e) {
         assertThat(exReference.get(), instanceOf(IllegalArgumentException.class));
     }
 
+    public void testAssertAtLeastOnceWillLogAssertionErrorWhenNotResolved() throws Exception {
+        try (MockLog mockLog = MockLog.capture(ExceptionsHelper.class)) {
+            mockLog.addExpectation(
+                new MockLog.PatternSeenEventExpectation(
+                    "action listener not called",
+                    ExceptionsHelper.class.getName(),
+                    Level.ERROR,
+                    ".*Expected listener NoopActionListener to be called at least once, but it was never called\\. Created:(.*|\\s)*"
+                )
+            );
+            final AtomicReference<ActionListener<Object>> listenerRef = new AtomicReference<>(
+                ActionListener.assertAtLeastOnce(ActionListener.noop())
+            );
+            assertBusy(() -> {
+                // Nullify reference so it becomes unreachable
+                listenerRef.set(null);
```

Comment:
Does setting the ref to null repeatedly make any difference?

Reply:
Nah I guess it was just to separate the creation of the reference and clearing of the reference into separate blocks.

Reply:
I moved it out to make it clearer what might need repeating